### PR TITLE
*: use the newly designed memdb

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
This is the 4th patchset of #18650. Please see the full description in the original PR page.

Now that the new memdb implementation is complete, we can replace the existing implementation with the new one. To avoid name conflicts, the code of the old memdb was completely deleted.

This patchset is based on https://github.com/pingcap/tidb/pull/18656, I'll rebase this branch after the parent PR is merged.

### Release note 

-  Replace transaction states buffer with a more optimized version.

### CODE REVIEW

**Comment by bobotu** [comment]
PTAL @jackysp @tiancaiamao @lysu @coocood

File: kv/sandbox.go [hunk 1]
```diff
@@ -98,7 +97,7 @@ func (sb *sandbox) PutWithFlags(key []byte, flags KeyFlags, value []byte) {
 	if !exists {
 		height = sb.randomHeight()
 	} else {
-		flags = flags.Merge(next[0].flags)
+		// flags = flags.Merge(next[0].flags)
```
Comment (coocood) [inline kv/sandbox.go:100]: Can we remove the commetted code?

File: kv/sandbox.go [hunk 2]
```diff
@@ -404,7 +403,7 @@ func (sb *sandbox) merge(new *sandbox) int {
 
 		height := int(newNode.height)
 		if exists {
-			newNode.flags = newNode.flags.Merge(ms.next[0].flags)
+			// newNode.flags = newNode.flags.Merge(ms.next[0].flags)
```
Comment (coocood) [inline kv/sandbox.go:406]: ditto

File: session/txn.go [hunk 3]
```diff
@@ -510,28 +422,23 @@ func (s *session) HasDirtyContent(tid int64) bool {
...
-		return err
-	}
+	st.flushStmtBuf()
 
-	var err error
 	failpoint.Inject("mockStmtCommitError", func(val failpoint.Value) {
-		if val.(bool) && bufLen > 3 {
-			err = errors.New("mock stmt commit error")
+		if val.(bool) && stmtLen > 3 {
+			err := errors.New("mock stmt commit error")
```
Comment (coocood) [inline session/txn.go:437]: Seems the only returned error is failpoint, can we remove the returned error for `StmtCommit`?

**Comment by coocood** [comment]
LGTM

File: session/txn.go [hunk 4]
```diff
@@ -386,20 +288,19 @@ func (st *TxnState) cleanup() {
...
+			return
 		}
 		// If the key is already locked, it will be deduplicated in LockKeys method later.
 		keys = append(keys, k)
-		return nil
-	}); err != nil {
-		return nil, err
-	}
+	})
+	// LockKeys will sort and deduplicate keys.
```
Comment (jackysp) [inline session/txn.go:303]: why add this comment?

Comment (bobotu) [reply session/txn.go:303]: Keys can only be returned from InspectStage according to the order of writing.

Comment (bobotu) [reply session/txn.go:303]: `LockKeys` already handled this in

https://github.com/pingcap/tidb/blob/6cf50d3f898cadf3e74adaa6c12ab0e671bd9104/store/tikv/txn.go#L437-L448

Comment (jackysp) [reply session/txn.go:303]: Yes, it already has a comment,
https://github.com/pingcap/tidb/pull/18657/files#diff-417f1f41cb3111fe7a10aa64ea77d1efR300

**Comment by jackysp** [comment]
PTAL @lysu

**Pull request merged**